internal/internal_workflow.go

```diff
@@ -1742,3 +1754,56 @@ func (us updateSchedulerImpl) Spawn(ctx Context, name string, highPriority bool,
 func (us updateSchedulerImpl) Yield(ctx Context, reason string) {
 	getState(ctx).yield(reason)
 }
+
+func (m *mutexImpl) Lock(ctx Context) error {
+	err := Await(ctx, func() bool {
+		return !m.locked
+	})
+	if err != nil {
+		return err
+	}
+	m.locked = true
+	return nil
+}
+
+func (m *mutexImpl) TryLock(ctx Context) bool {
+	if m.locked {
+		return false
+	}
+	m.locked = true
+	return true
+}
+
+func (m *mutexImpl) Unlock() {
+	if !m.locked {
+		panic("Mutex.Unlock() was called on an unlocked mutex")
+	}
+	m.locked = false
+}
+
+func (m *mutexImpl) IsLocked() bool {
+	return m.locked
+}
+
+func (s *semaphoreImpl) Acquire(ctx Context, n int64) error {
+	err := Await(ctx, func() bool {
+		return s.count >= n
+	})
+	if err != nil {
+		return err
+	}
+	s.count -= n
+	return nil
+}
+
+func (s *semaphoreImpl) TryAcquire(ctx Context, n int64) bool {
+	if s.count < n {
+		return false
+	}
+	s.count -= n
+	return true
+}
+
+func (s *semaphoreImpl) Release(n int64) {
+	s.count += n
```

Comment:
Arguably could have a panic here if count exceeds the initial number given at construction

Reply:
Let me double check what Go's semaphore does, I don't think it cares

Reply:
AH yes it should panic

Reply:
Added a panic here